core/src/typecheck/mod.rs

```diff
@@ -2224,7 +2284,10 @@ fn infer<L: Linearizer>(
                 .cloned()
                 .ok_or(TypecheckError::UnboundIdentifier(*x, *pos))?;
 
-            linearizer.add_term(lin, rt, x_ty.clone());
+            if let Some(ref id) = item_id {
+                linearizer.retype(lin, id, x_ty.clone());
+            }
```

Comment:
The amount of `if let Some(...)` in this function makes we wonder if it wouldn't make more sense to have `retype` operate on `Option<ItemId>`? Or, if that feels iffy semantically, we could have a wrapper that does that.

Reply:
That's a good idea. `retype` is only ever called from the typechecker, and the linearizer will be scrapped ultimately, so it's probably not very rewarding to try to keep the `retype` signature more "natural" and write a second wrapper somewhere else, so I'll probably just change the signature of `retype`.

Reply:
Addressed in [f39e49b](https://github.com/tweag/nickel/pull/1583/commits/f39e49b5571473d2208b69cb33dd4efd3d31ef6b)